### PR TITLE
Allow forward functions with single output to return Variable

### PR DESCRIPTION
**Pull request by malvika2147**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23803 Allow forward functions with single output to return Variable**

Summary:
Custom `forward()` can return a `Variable` in case of single outputs instead of returning a `variable_list` of size 1. 

Test Plan: Modified tests involving single output forward functions.

Differential Revision: [D16673857](https://our.internmc.facebook.com/intern/diff/D16673857)

### CODE REVIEW

File: torch/csrc/autograd/custom_function.h [hunk 1]
```diff
@@ -56,10 +56,13 @@ TORCH_API variable_list _wrap_outputs(
 // auto y = MyFunction::apply(6, x);
 // Example backward call:
 // y[0].sum().backward();
+template<typename X, typename... Args>
+using forward_t = decltype(X::forward(nullptr, std::declval<Args>()...));
```
Comment (ezyang) [inline torch/csrc/autograd/custom_function.h]: You probably didn't mean to attach the docblock above to `forward_t`. (Speaking of which, would be nice to have a doc for `forward_t` too!)

**Comment by ezyang** [comment]
@smessmer could you please help review the C++ template metaprogramming in this diff?

**Review by ezyang — APPROVED** [review]
Seems to work, but I don't know too much about how the template metaprogramming works here.

File: torch/csrc/autograd/custom_function.h [hunk 2]
```diff
@@ -182,7 +198,8 @@ variable_list Function<T>::apply(Args&&... args) {
       node->input_info_.emplace_back(var);
   }
 
-  variable_list outputs;
+  typedef forward_t<X, Args...> forward_return_t;
```
Comment (smessmer) [inline torch/csrc/autograd/custom_function.h]: nit: `using` is more readable than `typedef`

File: torch/csrc/autograd/custom_function.h [hunk 3]
```diff
@@ -159,9 +169,15 @@ template <typename... Args>
 void extract_vars(std::vector<bool> &is_var, variable_list& list, Args&& ... args) {
 }
 
+template <typename T>
+typename std::enable_if<std::is_same<T, variable_list>::value, T&>::type to_output_type(variable_list& output_list) { return output_list; }
```
Comment (smessmer) [inline torch/csrc/autograd/custom_function.h]: nit: we have `guts::enable_if_t` to avoid the `typename`

**Review by smessmer — COMMENTED** [review]
metaprogramming looks good

**Pull request closed without merging**